### PR TITLE
Teams request access dialog

### PR DESCRIPTION
**Pull request by buoyad**
Adds the join teams dialog and RPC call

@keybase/react-hackers

### CODE REVIEW

File: shared/teams/container.js [hunk 1]
```diff
@@ -52,8 +52,15 @@ const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
   },
   onHideBanner: () => dispatch(injectItem('sawChatBanner', 'true')),
   onJoinTeam: () => {
-    // TODO: Hook this up once we have a join team dialog.
-    console.log('onJoinTeam not implemented yet')
+    dispatch(
+      navigateAppend([
+        {
```
Comment (chrisnojima) [inline shared/teams/container.js]: i think if you're not passing props you can just do ['showJoinTeamDialog']

File: shared/teams/join-team/container.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// @flow
+import JoinTeamDialog from './'
+import {connect} from 'react-redux'
+import {compose, lifecycle, withState, withHandlers} from 'recompose'
+import {joinTeam, setTeamJoinError, setTeamJoinSuccess} from '../../actions/teams/creators'
+import {upperFirst} from 'lodash'
+
+import type {TypedState} from '../../constants/reducer'
+
+// TODO successText => successHeader; errorText => errorHeader
```
Comment (chrisnojima) [inline shared/teams/join-team/container.js]: 🔥

File: shared/teams/join-team/container.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// @flow
+import JoinTeamDialog from './'
+import {connect} from 'react-redux'
+import {compose, lifecycle, withState, withHandlers} from 'recompose'
+import {joinTeam, setTeamJoinError, setTeamJoinSuccess} from '../../actions/teams/creators'
+import {upperFirst} from 'lodash'
```
Comment (chrisnojima) [inline shared/teams/join-team/container.js]: prefer import upperFirst from 'lodash/upperFirst' so we don't pull in the entirety of lodash

File: shared/teams/join-team/index.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
+// @flow
+import React from 'react'
+import {Box, Button, HeaderHoc, Icon, Input, PopupDialog, Text, ScrollView} from '../../common-adapters/index'
+import {isMobile} from '../../constants/platform'
+import {globalColors, globalMargins, globalStyles} from '../../styles'
+
+import type {Props} from './'
```
Comment (chrisnojima) [inline shared/teams/join-team/index.js]: can be '.'

File: shared/teams/join-team/index.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+const styleContainer = {
+  ...globalStyles.flexBoxCenter,
+  ...(isMobile ? {} : {cursor: 'default'}),
```
Comment (chrisnojima) [inline shared/teams/join-team/index.js]: why is this needed?

File: shared/teams/join-team/index.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      paddingTop: globalMargins.xlarge,
+    }
+  : {
+      marginBottom: 80,
+      marginLeft: 80,
+      marginRight: 80,
+      marginTop: 90,
+    }
+
+const Contents = (props: Props) => (
```
Comment (chrisnojima) [inline shared/teams/join-team/index.js]: likely not needed just move ScrollView into renderContents.
Also if you're going to call a function, just make it a stateless component
rename renderContents -> RenderContents and call it

```
<ScrollView>
  <RenderContents {...props} />
</ScrollView>
```

(but really just get rid of Contents)

File: shared/teams/join-team/index.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  <ScrollView>
+    {renderContents(props)}
+  </ScrollView>
+)
+
+const renderContents = ({errorText, success, name, onNameChange, onSubmit, onBack}) => {
```
Comment (chrisnojima) [inline shared/teams/join-team/index.js]: better to split this into a Success and Entry components instead of this big if block

**Comment by buoyad** [comment]
This is using the wrong RPC, will fix in a moment

File: shared/teams/join-team/container.js (hunk 2 above)
Comment (buoyad) [reply shared/teams/join-team/container.js]: whoops, thought I got rid of that

File: shared/teams/join-team/index.js (hunk 5 above)
Comment (buoyad) [reply shared/teams/join-team/index.js]: it's not, I don't think. The basic styling was based off of the new-team dialog, which [includes this](https://github.com/keybase/client/blob/master/shared/teams/new-team/index.js#L59). I wasn't sure what was safe to remove, but it looks like it's dead code here.

File: shared/common-adapters/header-hoc.native.js [hunk 8]
```diff
@@ -10,10 +10,17 @@ import type {Props} from './header-hoc'
...
+    const {onBack, onCancel, headerStyle, title, customComponent, theme = 'light', showBorder = true} = props
     return (
       <Box style={_containerStyle}>
-        <Box style={{..._headerStyle, ..._headerStyleThemed[theme], ...headerStyle}}>
+        <Box
+          style={{
+            ..._headerStyle,
+            ..._headerStyleThemed[theme],
+            ...headerStyle,
+            borderBottomWidth: showBorder ? StyleSheet.hairlineWidth : 0,
```
Comment (chrisnojima) [inline shared/common-adapters/header-hoc.native.js]: this might not be safe as other components likely did this with the headerStyle prop instead, lets discuss

File: shared/teams/join-team/index.native.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      <Box style={{...globalStyles.flexBoxRow, marginTop: globalMargins.medium}}>
+        <Button type="Primary" style={{marginLeft: globalMargins.tiny}} onClick={onSubmit} label="Continue" />
+      </Box>
+    </Box>
+  </Box>
+)
+
+export const Contents = (props: Props) => (
+  <ScrollView>
+    {!!props.success && SuccessComponent(props)}
```
Comment (chrisnojima) [inline shared/teams/join-team/index.native.js]: if you're using compose already you can use branch for this kind of logic
also use <SuccessComponent {...props}/> vs calling a function

File: shared/teams/join-team/index.native.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    {!props.success && EntryComponent(props)}
+  </ScrollView>
+)
+
+export default compose(
+  withProps(props => ({
+    // customComponent: <Header {...props} />,
```
Comment (chrisnojima) [inline shared/teams/join-team/index.native.js]: 🔥

File: shared/common-adapters/header-hoc.native.js (hunk 8 above)
Comment (buoyad) [reply shared/common-adapters/header-hoc.native.js]: oh I completely missed that L20 didn't have an underscore, that makes more sense

File: shared/teams/join-team/index.desktop.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          />
+        </Box>
+      </Box>
+    </Box>
+  </ScrollView>
+)
+
+const PopupWrapped = (props: Props) => (
+  <PopupDialog onClose={props.onBack}>
+    {!!props.success && <SuccessComponent {...props} />}
```
Comment (chrisnojima) [inline shared/teams/join-team/index.desktop.js]: 🐑 you should be able to do this
```
props.success ? <SuccessComponent {...props} /> : <EntryComponent {...props} />
```

Comment (chrisnojima) [reply shared/teams/join-team/index.desktop.js]: or use branch like in .native
